File: examples/python/spectator.py
#!/usr/bin/env python3

#####################################################################
# This script presents SPECTATOR mode. In SPECTATOR mode you play and
# your agent can learn from it.
# Configuration is loaded from "../../scenarios/<SCENARIO_NAME>.cfg" file.
#
# To see the scenario description go to
# https://vizdoom.farama.org/main/environments/default/
#####################################################################

import os
from argparse import ArgumentParser
from time import sleep

import vizdoom as vzd


DEFAULT_CONFIG = os.path.join(vzd.scenarios_path, "deathmatch.cfg")

if __name__ == "__main__":
    parser = ArgumentParser("ViZDoom example showing how to use SPECTATOR mode.")
    parser.add_argument(
        dest="config",
        default=DEFAULT_CONFIG,
        nargs="?",
        help="Path to the configuration file of the scenario."
        " Please see "
        "../../scenarios/*cfg for more scenarios.",
    )
    parser.add_argument(
        "-e",
        "--episodes",
        default=1,
        type=int,
        help="Number of episodes to play.",
    )
    parser.add_argument(
        "-e",
        "--episodes",
        default=1,
        type=int,
        help="Number of episodes to play.",
    )
    args = parser.parse_args()
    game = vzd.DoomGame()

    # Choose scenario config file.
    game.load_config(args.config)

    # Enables freelook (mouse look) in the engine.
    game.add_game_args("+freelook 1")

    # Increate screen resolution for better experience.
    game.set_screen_resolution(vzd.ScreenResolution.RES_800X600)

    # Enables spectator mode, so you can play. Sounds strange but it is the agent who is supposed to watch not you.
    game.set_window_visible(True)
    game.set_mode(vzd.Mode.SPECTATOR)

    game.init()

    for i in range(args.episodes):
        print(f"Episode #{i + 1}")

        game.new_episode()
        while not game.is_episode_finished():
            state = game.get_state()
            assert state is not None

            game.advance_action()
            last_action = game.get_last_action()
            reward = game.get_last_reward()

            print(
                f"State #{state.number} | "
                f"Game variables: {state.game_variables} | "
                f"Action: {last_action} | "
                f"Reward: {reward}"
            )

        print("Episode finished!")
        print("Total reward:", game.get_total_reward())
        print("************************")
        sleep(2.0)

    game.close()

File: examples/c++/Spectator.cpp
#include "ViZDoom.h"
#include <iostream>
#include <vector>

using namespace vizdoom;

int main(){

    std::cout << "\n\nSPECTATOR EXAMPLE\n\n";


    DoomGame *game = new DoomGame();

    // Choose scenario config file you wish to be watched by agent.
    // Don't load two configs cause the second will overwrite the first one.
    // Multiple config files are ok but combining these ones doesn't make much sense.

    game->loadConfig("../../scenarios/basic.cfg");
    //game->loadConfig("../../scenarios/deadly_corridor.cfg");
    //game->loadConfig("../../scenarios/deathmatch.cfg");
    //game->loadConfig("../../scenarios/defend_the_center.cfg");
    //game->loadConfig("../../scenarios/defend_the_line.cfg");
    //game->loadConfig("../../scenarios/health_gathering.cfg");
    //game->loadConfig("../../scenarios/my_way_home.cfg");
    //game->loadConfig("../../scenarios/predict_position.cfg");
    //game->loadConfig("../../scenarios/take_cover.cfg");

    game->setDoomGamePath("../../bin/freedoom2.wad");
    //game->setDoomGamePath("../../bin/doom2.wad");      // Not provided with environment due to licences.

    game->setScreenResolution(RES_640X480);

    // Enables spectator mode, so You can play and agent watch your actions.
    // You can only use the buttons selected as available.
    game->setMode(SPECTATOR);

    game->init();

    // Run this many episodes
    int episodes = 10;

    for (int i = 0; i < episodes; ++i) {

        std::cout << "Episode #" << i + 1 << "\n";

        // Starts a new episode. It is not needed right after init() but it doesn't cost much and the loop is nicer.
        game->newEpisode();

        while (!game->isEpisodeFinished()) {

            // Get the state.
            GameStatePtr state = game->getState();

            // Advances action - lets You play next game tic.
            game->advanceAction();

            // You can also advance a few tics at once.
            // game->advanceAction(4);

            // Get the last action performed by You.
            std::vector<double> lastAction = game->getLastAction();

            // And reward You get.
            double reward = game->getLastReward();

            std::cout << "State #" << state->number << "\n";
            std::cout << "Action made:";
            for(auto a: lastAction) std::cout << " " << a;
            std::cout << "\n";
            std::cout << "Action reward: " << reward <<"\n";
            std::cout << "=====================\n";

        }

        std::cout << "Episode finished.\n";
        std::cout << "Total reward: " << game->getTotalReward() << "\n";
        std::cout << "************************\n";
    }

    // It will be done automatically in destructor but after close You can init it again with different settings.
    game->close();
    delete game;
}
